### ISSUE TITLE
question for patroni auto failover issue

### ISSUE THREAD

**Issue by louteq** [zalando/patroni#1860]
**Describe the bug**
Hi team, we have thousands of postgres instance managed by patroni, so for such amount instances we always try to handle instance crash/machine shutdown... issue automatically.

But one scenario we found patroni can't auto failover, can't elet new leader though no any lag on standby. I saw one friend submitted same issue recently. the link is : https://github.com/zalando/patroni/issues/1858

I checked it but not find a solution , seems it is already closed.

under some circumstances  our postgres instance server will shutdown for many reasons, we found patroni can't elect new leader even though there's no any lag between primary and standby, so no data loss. under such cases, we hope patroni can elect new leader, if not , all instance will become replica and read only, our business will be impacted. And we have to check and handle manually, it's very time consuming.

but if we kill patroni process or kill -9 patroni process, new leader can select out.

After investigation, I guess it is caused as when machine shutdown , postgres will do a checkpoint and wal senders were teminated , so lead patroni checked wal lag behind one wal segment size.

but same time, when compared with kill patroni process, we found postgres also will do a checkpoint.

below is postgres log for above 2 scenarios:
*****kill patroni process, can elect new cluster leader*****
2021-02-26 16:41:15.519 CST,,,2246,,6038b139.8c6,3,,2021-02-26 16:28:41 CST,,0,LOG,00000,"received fast shutdown request",,,,,,,,,""
2021-02-26 16:41:15.520 CST,,,2246,,6038b139.8c6,4,,2021-02-26 16:28:41 CST,,0,LOG,00000,"aborting any active transactions",,,,,,,,,""
2021-02-26 16:41:15.520 CST,"postgres","postgres",2262,"127.0.0.1:49178",6038b13a.8d6,3,"idle",2021-02-26 16:28:42 CST,4/0,0,FATAL,57P01,"terminating connection due to administrator command",,,,,,,,,"Patroni"
2021-02-26 16:41:15.520 CST,"postgres","postgres",2262,"127.0.0.1:49178",6038b13a.8d6,4,"idle",2021-02-26 16:28:42 CST,,0,LOG,00000,"disconnection: session time: 0:12:32.596 user=postgres database=postgres host=127.0.0.1 port=49178",,,,,,,,,"Patroni"
2021-02-26 16:41:15.521 CST,,,2256,,6038b139.8d0,2,,2021-02-26 16:28:41 CST,1/0,0,LOG,00000,"bgworker pgsentinel signal: processed SIGTERM",,,,,,,,,""
2021-02-26 16:41:15.523 CST,,,2246,,6038b139.8c6,5,,2021-02-26 16:28:41 CST,,0,LOG,00000,"background worker ""logical replication launcher"" (PID 2257) exited with exit code 1",,,,,,,,,""
2021-02-26 16:41:15.523 CST,,,2250,,6038b139.8ca,7,,2021-02-26 16:28:41 CST,,0,LOG,00000,"shutting down",,,,,,,,,""
2021-02-26 16:41:15.581 CST,,,2250,,6038b139.8ca,8,,2021-02-26 16:28:41 CST,,0,LOG,00000,"checkpoint starting: shutdown immediate",,,,,,,,,""
2021-02-26 16:41:15.748 CST,,,2250,,6038b139.8ca,9,,2021-02-26 16:28:41 CST,,0,LOG,00000,"checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.003 s, sync=0.000 s, total=0.177 s; sync files=0, longest=0.000 s, average=0.000 s; distance=32767 kB, estimate=32767 kB",,,,,,,,,""
2021-02-26 16:41:15.925 CST,"repl","",2264,"192.168.56.95:41292",6038b140.8d8,3,"streaming 0/E0000A0",2021-02-26 16:28:48 CST,,0,LOG,00000,"disconnection: session time: 0:12:27.735 user=repl database= host=192.168.56.95 port=41292",,,,,,,,,"awsdb02"
2021-02-26 16:41:15.928 CST,"repl","",2263,"192.168.56.96:40288",6038b13e.8d7,3,"streaming 0/E0000A0",2021-02-26 16:28:46 CST,,0,LOG,00000,"disconnection: session time: 0:12:29.213 user=repl database= host=192.168.56.96 port=40288",,,,,,,,,"awsdb03"
2021-02-26 16:41:15.929 CST,,,2537,"192.168.56.95:41306",6038b42b.9e9,1,"",2021-02-26 16:41:15 CST,,0,LOG,00000,"connection received: host=192.168.56.95 port=41306",,,,,,,,,""
2021-02-26 16:41:15.929 CST,"repl","",2537,"192.168.56.95:41306",6038b42b.9e9,2,"",2021-02-26 16:41:15 CST,,0,FATAL,57P03,"the database system is shutting down",,,,,,,,,""
2021-02-26 16:41:15.939 CST,,,2246,,6038b139.8c6,6,,2021-02-26 16:28:41 CST,,0,LOG,00000,"database system is shut down",,,,,,,,,""

*****shutdown leader machine, can't elect new leader, lag one wal segment size*****
2021-02-26 16:42:54.049 CST,,,2213,,6038b138.8a5,4,,2021-02-26 16:28:40 CST,,0,LOG,00000,"received SIGHUP, reloading configuration files",,,,,,,,,""
2021-02-26 16:42:54.050 CST,,,2213,,6038b138.8a5,5,,2021-02-26 16:28:40 CST,,0,LOG,00000,"received smart shutdown request",,,,,,,,,""
2021-02-26 16:42:54.052 CST,,,2221,,6038b138.8ad,2,,2021-02-26 16:28:40 CST,2/0,0,LOG,00000,"bgworker pgsentinel signal: processed SIGHUP",,,,,,,,,""
2021-02-26 16:42:54.052 CST,,,2221,,6038b138.8ad,3,,2021-02-26 16:28:40 CST,2/0,0,LOG,00000,"bgworker pgsentinel signal: processed SIGTERM",,,,,,,,,""
2021-02-26 16:42:54.053 CST,"postgres","postgres",2228,"127.0.0.1:50292",6038b13a.8b4,3,"idle",2021-02-26 16:28:42 CST,3/0,0,FATAL,57P01,"terminating connection due to administrator command",,,,,,,,,"Patroni"
2021-02-26 16:42:54.053 CST,"postgres","postgres",2228,"127.0.0.1:50292",6038b13a.8b4,4,"idle",2021-02-26 16:28:42 CST,,0,LOG,00000,"disconnection: session time: 0:14:11.145 user=postgres database=postgres host=127.0.0.1 port=50292",,,,,,,,,"Patroni"
2021-02-26 16:42:54.054 CST,"repl","",2399,"192.168.56.95:60288",6038b431.95f,3,"streaming 0/E0002A0",2021-02-26 16:41:21 CST,5/0,0,FATAL,57P01,"terminating connection due to administrator command",,,,,,,,,"awsdb02"
2021-02-26 16:42:54.054 CST,"repl","",2399,"192.168.56.95:60288",6038b431.95f,4,"streaming 0/E0002A0",2021-02-26 16:41:21 CST,,0,LOG,00000,"disconnection: session time: 0:01:32.723 user=repl database= host=192.168.56.95 port=60288",,,,,,,,,"awsdb02"
2021-02-26 16:42:54.055 CST,"repl","",2414,"192.168.56.94:37980",6038b45d.96e,3,"streaming 0/E0002A0",2021-02-26 16:42:05 CST,6/0,0,FATAL,57P01,"terminating connection due to administrator command",,,,,,,,,"awsdb01"
2021-02-26 16:42:54.055 CST,"repl","",2414,"192.168.56.94:37980",6038b45d.96e,4,"streaming 0/E0002A0",2021-02-26 16:42:05 CST,,0,LOG,00000,"disconnection: session time: 0:00:48.465 user=repl database= host=192.168.56.94 port=37980",,,,,,,,,"awsdb01"
2021-02-26 16:42:54.072 CST,,,2213,,6038b138.8a5,6,,2021-02-26 16:28:40 CST,,0,LOG,00000,"background worker ""logical replication launcher"" (PID 2395) exited with exit code 1",,,,,,,,,""
2021-02-26 16:42:54.072 CST,,,2217,,6038b138.8a9,11,,2021-02-26 16:28:40 CST,,0,LOG,00000,"shutting down",,,,,,,,,""
2021-02-26 16:42:54.136 CST,,,2552,"192.168.56.94:38014",6038b48e.9f8,1,"",2021-02-26 16:42:54 CST,,0,LOG,00000,"connection received: host=192.168.56.94 port=38014",,,,,,,,,""
2021-02-26 16:42:54.136 CST,"repl","",2552,"192.168.56.94:38014",6038b48e.9f8,2,"",2021-02-26 16:42:54 CST,,0,FATAL,57P03,"the database system is shutting down",,,,,,,,,""
2021-02-26 16:42:54.148 CST,,,2553,"192.168.56.95:60304",6038b48e.9f9,1,"",2021-02-26 16:42:54 CST,,0,LOG,00000,"connection received: host=192.168.56.95 port=60304",,,,,,,,,""
2021-02-26 16:42:54.148 CST,"repl","",2553,"192.168.56.95:60304",6038b48e.9f9,2,"",2021-02-26 16:42:54 CST,,0,FATAL,57P03,"the database system is shutting down",,,,,,,,,""
2021-02-26 16:42:54.185 CST,,,2217,,6038b138.8a9,12,,2021-02-26 16:28:40 CST,,0,LOG,00000,"checkpoint starting: shutdown immediate",,,,,,,,,""
2021-02-26 16:42:54.190 CST,,,2213,,6038b138.8a5,7,,2021-02-26 16:28:40 CST,,0,LOG,00000,"received fast shutdown request",,,,,,,,,""
2021-02-26 16:42:54.345 CST,,,2217,,6038b138.8a9,13,,2021-02-26 16:28:40 CST,,0,LOG,00000,"checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.001 s, sync=0.000 s, total=0.166 s; sync files=0, longest=0.000 s, average=0.000 s; distance=32767 kB, estimate=32767 kB",,,,,,,,,""
2021-02-26 16:42:54.359 CST,,,2213,,6038b138.8a5,8,,2021-02-26 16:28:40 CST,,0,LOG,00000,"database system is shut down",,,,,,,,,""

Both did checkpoint. And currently to avoid handle machine shutdown issue manually , we can only set maximum_lag_on_failover high than wal_segment_size. but for important db, set the option value two high is dangeous if lag do exist.

could you please help on this and provice me a solution? let patroni elect new leader and avoid manual handle .

Thanks in advance.

**To Reproduce**

**Expected behavior**

**Screenshots**

**Environment**
- Patroni version: 2.0.1
- PostgreSQL version: 12.3
- DCS (and its version): ectd 3.3.4

**Patroni configuration file**
```
Please copy&paste your Patroni configuration file here
```

**patronictl show-config**
```
Please copy&paste the output of "patronictl show-config" command here
```

**Have you checked Patroni logs?**
Please provide a snippet of Patroni log files here

**Have you checked PostgreSQL logs?**
Please provide a snippet here

**Have you tried to use GitHub issue search?**
Maybe there is already a similar issue solved.

**Additional context**

**Comment by CyberDem0n** [comment]
We are running in circles.

The issue you are referring to was closed because there is no problem with Patroni.
Patroni does everything correctly. It knows that the former primary was on the LSN=X when it was stopped, but replicas are on LSN=Y. The difference between X and Y clearly exceeds the `maximum_lag_on_failover`, therefore the failover is not triggered. X and Y are not taken arbitrarily by Patroni, these are metrics reported by Postgres, Patroni can't ignore them.

I also explained many times why this difference becomes so big. The reason is that the walsender processes on the primary are prematurely terminated by your init system, therefore replicas are cleary missing some amount of WAL.

**Comment by louteq** [comment]
replica reported it's behind one segment file size lag. so replica calculated the lag based on the former leaders LSN X from DCS, and it's own LSN from DCS.

is it based on below "xlog_location\":603979776 ?

[root@localhost ~]# curl http://127.0.0.1:2379/v2/keys/?recursive=true | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2586    0  2586    0     0  3042k      0 --:--:-- --:--:-- --:--:-- 2525k
{
  "action": "get",
  "node": {
    "dir": true,
    "nodes": [
      {
        "key": "/service",
        "dir": true,
        "nodes": [
          {
            "key": "/service/awsdb",
            "dir": true,
            "nodes": [
              {
                "key": "/service/awsdb/members",
                "dir": true,
                "nodes": [
                  {
                    "key": "/service/awsdb/members/awsdb01",
                    "value": "{\"conn_url\":\"postgres://192.168.56.94:54322/postgres\",\"api_url\":\"http://192.168.56.94:8008/patroni\",\"state\":\"running\",\"role\":\"master\",\"version\":\"2.0.1\",\"xlog_location\":603979776,\"timeline\":5}",
                    "expiration": "2021-02-26T09:31:23.716206553Z",
                    "ttl": 20,
                    "modifiedIndex": 566862,
                    "createdIndex": 566862
                  },
                  {
                    "key": "/service/awsdb/members/awsdb03",
                    "value": "{\"conn_url\":\"postgres://192.168.56.96:54322/postgres\",\"api_url\":\"http://192.168.56.96:8008/patroni\",\"state\":\"running\",\"role\":\"replica\",\"version\":\"2.0.1\",\"xlog_location\":603979776,\"timeline\":5}",
                    "expiration": "2021-02-26T09:31:23.728861806Z",
                    "ttl": 20,
                    "modifiedIndex": 566863,
                    "createdIndex": 566863
                  },
                  {
                    "key": "/service/awsdb/members/awsdb02",
                    "value": "{\"conn_url\":\"postgres://192.168.56.95:54322/postgres\",\"api_url\":\"http://192.168.56.95:8008/patroni\",\"state\":\"running\",\"role\":\"replica\",\"version\":\"2.0.1\",\"xlog_location\":603979776,\"timeline\":5}",
                    "expiration": "2021-02-26T09:31:23.728409918Z",
                    "ttl": 20,
                    "modifiedIndex": 566864,
                    "createdIndex": 566864
                  }
                ],
                "modifiedIndex": 564193,
                "createdIndex": 564193
              },
              {
                "key": "/service/awsdb/initialize",
                "value": "6933485761978124367",
                "modifiedIndex": 564198,
                "createdIndex": 564194
              },
              {
                "key": "/service/awsdb/config",
                "value": "{\"ttl\":20,\"loop_wait\":5,\"retry_timeout\":10,\"maximum_lag_on_failover\":1048576,\"postgresql\":{\"use_pg_rewind\":true,\"remove_data_directory_on_rewind_failure\":true,\"parameters\":null}}",
                "modifiedIndex": 564199,
                "createdIndex": 564199
              },
              {
                "key": "/service/awsdb/leader",
                "value": "awsdb01",
                "expiration": "2021-02-26T09:31:23.710866839Z",
                "ttl": 20,
                "modifiedIndex": 566861,
                "createdIndex": 566504
              },
              {
                "key": "/service/awsdb/optime",
                "dir": true,
                "nodes": [
                  {
                    "key": "/service/awsdb/optime/leader",
                    "value": "603979776",
                    "modifiedIndex": 566797,
                    "createdIndex": 566797
                  }
                ],
                "modifiedIndex": 564203,
                "createdIndex": 564203
              },
              {
                "key": "/service/awsdb/history",
                "value": "[[1,234881184,\"no recovery target specified\",\"2021-02-26T16:41:16.115020+08:00\"],[2,268435616,\"no recovery target specified\",\"2021-02-26T17:04:43.986605+08:00\"],[3,402653184,\"no recovery target specified\",\"2021-02-26T17:15:27.203267+08:00\"],[4,503316480,\"no recovery target specified\",\"2021-02-26T17:23:42.727003+08:00\"]]",
                "modifiedIndex": 566509,
                "createdIndex": 566509
              }
            ],
            "modifiedIndex": 564193,
            "createdIndex": 564193
          }
        ],
        "modifiedIndex": 9,
        "createdIndex": 9
      }
    ]
  }
}
[root@localhost ~]#

**Comment by louteq** [comment]
can patroni do a enhancement , skip the latest shutdown db LSN. As replica is streaming , and really no lag and no data loss.

can patroni identify this and skip the latest shutdown DB LSN, and let cluster elect new leader？

referred the  #1858,  I also checked and found when I set archive_timeout=0, always can't select new leader? but if set archive_timeout to a non-zero value, it happen occasionally.

**Comment by CyberDem0n** [comment]
> can patroni do a enhancement , skip the latest shutdown db LSN.

Yes, probably. It would require running pg_waldump though.

> As replica is streaming , and really no lag and no data loss.

Well, the issue is that **realicas are not streaming** during the shutdown, because walsender processes are killed!

> I set archive_timeout=0, always can't select new leader? but if set archive_timeout to a non-zero value, it happen occasionally.

I believe the archiver process suffers from the same issue as walsender, i.e. it is being prematurely terminated. Depending on value set there are different chances of having some files archived or not.

**Comment by louteq** [comment]
> > can patroni do a enhancement , skip the latest shutdown db LSN.
> 
> Yes, probably. It would require running pg_waldump though.
> 
> > As replica is streaming , and really no lag and no data loss.
> 
> Well, the issue is that **realicas are not streaming** during the shutdown, because walsender processes are killed!
> 
> > I set archive_timeout=0, always can't select new leader? but if set archive_timeout to a non-zero value, it happen occasionally.
> 
> I believe the archiver process suffers from the same issue as walsender, i.e. it is being prematurely terminated. Depending on value set there are different chances of having some files archived or not.

thanks for your response, hope this can be handled smartly by patroni. I'm a Postgres DBA maintain lots of instances,  we meet such cases many times and spent lots of time to handle this manually ,if the enhancement can be archived, it will help me a lot.

**Comment by louteq** [comment]
Hi , so this will be fixed in next release 2.0.3, is that?

**Comment by CyberDem0n** [comment]
@louteq I would recommend you to apply the patch and test it.
